### PR TITLE
mvcc: check if already rolled back when prewrite.

### PR DESCRIPTION
**Pull request by disksing**
@siddontang @zhangjinpeng1987 

The bug is mentioned in #921

### CODE REVIEW

**Comment by ngaut** [comment]
Any tests?

**Comment by disksing** [comment]
@ngaut This PR only checked we cannot retry _Prewrite_ after _Rollback_. https://github.com/pingcap/tikv/pull/941/files#diff-10ac7ebd80ca4da7cccc8e3539f12eb4R214

The situation described in the issue can be simulated on TiDB with following code (which will be included in later PR on TiDB):
```
func (s *testCommitterSuite) TestPrewriteRollback(c *C) {
	s.mustCommit(c, map[string]string{
		"a": "a0",
		"b": "b0",
	})

	txn1 := s.begin(c)
	err := txn1.Set([]byte("a"), []byte("a1"))
	c.Assert(err, IsNil)
	err = txn1.Set([]byte("b"), []byte("b1"))
	c.Assert(err, IsNil)
	committer, err := newTxnCommitter(txn1)
	c.Assert(err, IsNil)
	err = committer.prewriteKeys(NewBackoffer(prewriteMaxBackoff), committer.keys)
	c.Assert(err, IsNil)

	txn2 := s.begin(c)
	v, err := txn2.Get([]byte("a"))
	c.Assert(v, BytesEquals, []byte("a0"))

	err = committer.prewriteKeys(NewBackoffer(prewriteMaxBackoff), committer.keys)
	c.Assert(err, IsNil)
	committer.commitTS, err = s.store.oracle.GetTimestamp()
	c.Assert(err, IsNil)
	err = committer.commitKeys(NewBackoffer(commitMaxBackoff), [][]byte{[]byte("a")})
	c.Assert(err, IsNil)

	txn3 := s.begin(c)
	v, err = txn3.Get([]byte("b"))
	c.Assert(err, IsNil)
	c.Assert(v, BytesEquals, []byte("b1"))
}
```

**Comment by siddontang** [comment]
can we simulate ticlient in tikv, so we can test these in tikv directly later.

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -59,7 +59,7 @@ impl<'a> MvccTxn<'a> {
 
     pub fn prewrite(&mut self, mutation: Mutation, primary: &[u8]) -> Result<()> {
         let key = mutation.key();
-        if let Some((_, commit)) = try!(self.reader.seek_write(&key, u64::max_value())) {
+        if let Some((commit, _)) = try!(self.reader.seek_write(&key, u64::max_value())) {
```
Comment (siddontang) [inline src/storage/mvcc/txn.rs]: no need to check rollback type?

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -125,27 +131,33 @@ impl<'a> MvccTxn<'a> {
...
             if !after_safe_point {
                 if commit <= safe_point {
                     // Set `after_safe_point` after the latest write after `safe_point`.
                     after_safe_point = true;
-                    // Latest write can be deleted if it's a `Modify::Delete`.
-                    if try!(self.reader.load_data(key, start)).is_none() {
-                        self.writes.push(Modify::Delete(CF_WRITE, key.append_ts(commit)));
+                    // Latest write can be deleted if its type is Delete/Rollback.
+                    match write.write_type {
+                        WriteType::Delete | WriteType::Rollback => {
```
Comment (siddontang) [inline src/storage/mvcc/txn.rs]: may still meet a slow prewrite after gc rollback. we should check in commit later.

File: src/storage/mvcc/txn.rs (hunk 1 above)
Comment (disksing) [inline src/storage/mvcc/txn.rs]: No. Here we can simply reject the _Prewrite_ no matter old write is _Put_, _Delete_, or _Rollback_.

**Comment by disksing** [comment]
@siddontang 
That would be a lot of work to mock tikv client...
I think for this issue it's okay to test like this. 'Prewrite after Rollback result in OK' is already not the behavior that we expected.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**